python-lib/ontology_tagging/formatter_by_document.py

```diff
@@ -177,86 +169,34 @@ def _write_row_category(
         """
         language = self._get_document_language(row, language_column)
         document_to_match = self._get_document_to_match(row, language)
-        matched_sentence, keyword_list = [], []
-        tag_columns_for_json, line, line_full = (
-            defaultdict(),
-            defaultdict(list),
-            defaultdict(defaultdict),
-        )
+        matched_sentence = []
+        keyword_list = []
+        categories_and_tags = defaultdict(list)
         for idx, sentence in enumerate(document_to_match):
-            original_sentence = row[self.text_column_tokenized]
+            original_sentence = row[self.text_column_tokenized][idx]
             for keyword in sentence.ents:
-                line, line_full = self._get_tags_in_row_category(
-                    match=keyword,
-                    line=line,
-                    line_full=line_full,
-                    sentence=original_sentence[idx],
-                    language=language,
-                )
-                keyword_list.append(keyword.text + " ")
-                matched_sentence.append(original_sentence[idx] + "\n")
-            tag_columns_for_json["tag_json_categories"] = self._fill_tags(
-                condition=(line and one_row_per_doc_json), value=dict(line)
-            )
-            tag_columns_for_json["tag_json_full"] = self._fill_tags(
-                condition=(line and one_row_per_doc_json),
-                value={
-                    column_name: dict(value) for column_name, value in line_full.items()
-                },
-            )
-        line = {column: self._fill_tags(True, line[column]) for column in line}
+                tag = keyword.ent_id_
+                category = keyword.label_
+                if tag not in categories_and_tags[category]:
+                    categories_and_tags[category].append(tag)
+                if keyword.text not in keyword_list:
+                    keyword_list.append(keyword.text)
```

Comment:
Nitpick: You could use sets instead of list for this operation. This way you don't need to check if element is in list before appending it, you simple you the set's `add` method.
Finally if you need them as a list, you can convert them in the end.

This could be a bad idea if you need to keep the order of appearance, as python sets don't guarantee to keep this: https://stackoverflow.com/questions/61414947/why-dont-python-sets-preserve-insertion-order

Reply:
Yes, set is really practical for this ! In this case, I do need to keep the order for debugging purposes (e.g to compare easily the matched sentences / the found keywords)